### PR TITLE
Solutions - module04

### PR DESCRIPTION
**Pull request by sstoyanov7**
Solutions

- [x] 01-Cash-Desk
- [x] 02-Bank-Account
- [x] 03-Panda-Social-Network

### CODE REVIEW

**Review by zahari — APPROVED** [review]
Everything looks good in CashDesk

File: module04/01-Cash-Desk/solution.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+class Bill:
```
Comment (zahari) [inline module04/01-Cash-Desk/solution.py]: You can try creating classes that inherit from `object` e.g. `SomeExample(object):` this is the new style class

File: module04/01-Cash-Desk/solution.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        if self.__int__() == other.__int__():
+            return True
+        else:
+            return False
+
+    def __hash__(self):
+        return hash(self.amount)
+
+
+class BatchBill:
```
Comment (zahari) [inline module04/01-Cash-Desk/solution.py]: Same as above

File: module04/01-Cash-Desk/solution.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        s = 0
+        for bill in self.bills:
+            s += bill.amount
+        return s
+
+    def __getitem__(self, index):
+        return self.bills[index]
+
+
+class CashDesk:
```
Comment (zahari) [inline module04/01-Cash-Desk/solution.py]: Same as above

File: module04/01-Cash-Desk/solution.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    def inspect(self):
+        s = 'We have a total of ' + str(self.total()) + '$ in the desk' + '\n'
+        s += 'We have the following count of bills, sorted in ascending order:'
+        keys = sorted(self.banknotes.keys())
+        for k in keys:
+            s += '\n' + str(k) + '$ bills - ' + str(self.banknotes[k])
+        return s
+
+
+
```
Comment (zahari) [inline module04/01-Cash-Desk/solution.py:87]: Please use `__main__` block for the extra code that you use to test for yourself while developing

File: module04/03-Panda-Social-Network/solution.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        if self.has_panda(panda):
+            raise ValueError(f'I\'s already added {str(panda)} to the network.')
+
+        key = panda.__hash__()
+        self.network[key] = {key: panda}
+
+        return
+
+    def has_panda(self, panda):
+        key = panda.__hash__()
```
Comment (zahari) [inline module04/03-Panda-Social-Network/solution.py]: You can simply use `hash(panda)` no actual need to use the system-defined one of the class that one is not meant to call.

File: module04/03-Panda-Social-Network/solution.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    def make_friends(self, panda1, panda2):
+        if (not self.has_panda(panda1)):
+            self.add_panda(panda1)
+
+        if (not self.has_panda(panda2)):
+            self.add_panda(panda2)
+
+        key1 = panda1.__hash__()
+        key2 = panda2.__hash__()
```
Comment (zahari) [inline module04/03-Panda-Social-Network/solution.py]: Same as above

File: module04/03-Panda-Social-Network/solution.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        self.network[key1][key2] = panda2
+        self.network[key2][key1] = panda1
+
+        return
+
+    def are_friends(self, panda1, panda2):
+        key1 = panda1.__hash__()
+        key2 = panda2.__hash__()
```
Comment (zahari) [inline module04/03-Panda-Social-Network/solution.py]: Same as above

File: module04/03-Panda-Social-Network/solution.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    def __init__(self):
+        self.network = {}
+        self.friends = {}
+        return
+
+    def add_panda(self, panda):
+        if self.has_panda(panda):
+            raise ValueError(f'I\'s already added {str(panda)} to the network.')
+
+        key = panda.__hash__()
```
Comment (zahari) [inline module04/03-Panda-Social-Network/solution.py]: Same as below

File: module04/03-Panda-Social-Network/solution.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+    def __eq__(self, other):
+        result = self._name == other.name() and \
+                 self._email == other.email() and \
+                 self._gender == other.gender()
+
+        return result
+
+    def __hash__(self):
+        s = self.__str__()
```
Comment (zahari) [inline module04/03-Panda-Social-Network/solution.py]: You can most likely call `str(self)`

File: module04/03-Panda-Social-Network/solution.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
+class Panda:
```
Comment (zahari) [inline module04/03-Panda-Social-Network/solution.py]: You can use new-style class that inherits from `object`

File: module04/03-Panda-Social-Network/solution.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        s = self.__str__()
+        h = hash(s)
+        h2 = hash((self.name(), self.email(), self.gender()))
+        return h
+
+    def __str__(self):
+        return f'Panda: ({self.name()}, {self.email()}, {self.gender()})'
+
+
+class PandaSocialNetwork:
```
Comment (zahari) [inline module04/03-Panda-Social-Network/solution.py]: Same as above

File: module04/01-Cash-Desk/solution.py (hunk 4 above)
Comment (sstoyanov7) [reply module04/01-Cash-Desk/solution.py:87]: Good point!

**Comment by sstoyanov7** [comment]
@zahari , addressed the comments.

**Pull request merged**